OrbitLinuxTracing/TracerThread.cpp

```diff
@@ -322,6 +216,174 @@ void TracerThread::Run(
+  uint16_t cpu = static_cast<uint16_t>(event.GetCpu());
+  uint64_t time = event.GetTimestamp();
+
+  if (event.IsSwitchOut()) {
+    listener_->OnContextSwitchOut(ContextSwitchOut(tid, cpu, time));
+  } else {
+    listener_->OnContextSwitchIn(ContextSwitchIn(tid, cpu, time));
+  }
+}
+
+void TracerThread::ProcessContextSwitchCpuWideEvent(
+    const perf_event_header& header, PerfEventRingBuffer& ring_buffer) {
+  SystemWideContextSwitchPerfEvent event;
+  ring_buffer.ConsumeRecord(header, &event.ring_buffer_record);
+
+  if (event.GetPrevTid() != 0) {
+    ContextSwitchOut context_switch_out{event.GetPrevTid(),
+                                        static_cast<uint16_t>(event.GetCpu()),
+                                        event.GetTimestamp()};
+    listener_->OnContextSwitchOut(context_switch_out);
+  }
+  if (event.GetNextTid() != 0) {
+    ContextSwitchIn context_switch_in{event.GetNextTid(),
+                                      static_cast<uint16_t>(event.GetCpu()),
+                                      event.GetTimestamp()};
+    listener_->OnContextSwitchIn(context_switch_in);
+  }
+
+  ++sched_switch_count_;
+}
+
+void TracerThread::ProcessForkEvent(const perf_event_header& header,
+                                    PerfEventRingBuffer& ring_buffer) {
+  ForkPerfEvent event;
+  ring_buffer.ConsumeRecord(header, &event.ring_buffer_record);
+
+  if (event.GetPid() == pid_) {
+    new_threads_.push_back(event.GetTid());
+  }
+}
+
+void TracerThread::ProcessExitEvent(const perf_event_header& header,
+                                    PerfEventRingBuffer& ring_buffer) {
+  ExitPerfEvent event;
+  ring_buffer.ConsumeRecord(header, &event.ring_buffer_record);
+
+  if (event.GetPid() == pid_) {
+    if (threads_to_fd_.count(event.GetTid()) > 0) {
+      int sample_fd = threads_to_fd_.at(event.GetTid());
+      perf_event_disable(sample_fd);
+      close(sample_fd);
+      // Do not remove the ring buffer from fds_to_ring_buffer here as
+      // we are already iterating over fds_to_ring_buffer.
+      fds_to_remove_.push_back(sample_fd);
+      threads_to_fd_.erase(sample_fd);
+    }
+  }
+}
+
+void TracerThread::ProcessMmapEvent(const perf_event_header& header,
+                                    PerfEventRingBuffer& ring_buffer) {
+  // There was a call to mmap with PROT_EXEC, hence refresh the maps.
+  // This should happen rarely.
+  ring_buffer.SkipRecord(header);
+  int fd = ring_buffer.GetFileDescriptor();
+  auto event =
+      std::make_unique<MapsPerfEvent>(MonotonicTimestampNs(), ReadMaps(pid_));
+  event->SetFileDescriptor(fd);
+  DeferEvent(std::move(event));
+}
+
+void TracerThread::ProcessSampleEvent(const perf_event_header& header,
+                                      PerfEventRingBuffer& ring_buffer) {
+  int fd = ring_buffer.GetFileDescriptor();
+  bool is_probe = uprobe_fds_to_function_.count(fd) > 0;
+  bool is_uprobe = is_probe && header.size > 32;  // TODO: find better test.
```

Comment:
Maybe instead of 32 we could use sizeof(perf_event_empty_sample), which should be the record that we get for uretprobes?

Reply:
Done, thanks!